liquidprompt

```diff
@@ -2361,17 +2387,42 @@ case "$LP_OS" in
         # First, remove the beginning of the line...
         lp_battery="${acpi#Battery *, }"
         lp_battery="${lp_battery%%%*}" # remove everything starting at '%'
+        lp_battery_status="${acpi}"
+    }
+
+    __lp_battery_detect() {
+        local lp_battery
+        unset _LP_BATTERY_FUNCTION
+        # First check SYSFS way
+        _LP_BATTERY_FUNCTION=__lp_battery_sysfs
+        "$_LP_BATTERY_FUNCTION" 2>/dev/null
+        [[ -n "${lp_battery-}" ]] && return 0
+        unset _LP_BATTERY_FUNCTION
+
+        # Try with ACPI
+        _LP_BATTERY_FUNCTION=__lp_battery_acpi
+        "$_LP_BATTERY_FUNCTION" 2>/dev/null
+        [[ -n "${lp_battery-}" ]] && return 0
+        unset _LP_BATTERY_FUNCTION
+        return 1
+    }
+
+    _lp_battery() {
+        (( LP_ENABLE_BATT )) || return 5
+
+        unset lp_battery lp_battery_status
+        "$_LP_BATTERY_FUNCTION"
 
         if [[ -z "${lp_battery}" ]]; then
```

Comment:
Since `lp_battery` is unset above, it needs to be guarded in the check here:
```suggestion
        if [[ -z "${lp_battery-}" ]]; then
```

Reply:
Absolutely makes sense, and to be honest, I'm a bit ashamed I missed the part of using specific files instead of the uevent one. I did come across some info about it, but can't remember where. Solution you proposed is definitely better, so I'll work on those changes. And thanks for educating me on fork() and memcopy, I still have A LOT to learn about shell programming.

Reply:
> And thanks for educating me on fork() and memcopy, I still have A LOT to learn about shell programming.

That's what I'm here for. I've spent way too much time on shell programing, and also on performance improvements for Liquidprompt. Keep the PRs coming, and I can verify that they don't hurt performance.

The rule of thumb though is don't call to any external binary unless you must. `grep`, `sed`, `cat`, and `awk` are all useful tools, but slow down the prompt. This advice often runs contrary to the best standards usually found for shell, but when you need to program with top performance in mind, you really need to cut corners.